src/features/goblins/merchant/lib/frogMachine.ts

```diff
@@ -0,0 +1,228 @@
+            },
+            {
+              target: "check_whitelist",
+              actions: assignFrogState,
+              cond: (_, event) => event.data.isWhitelistEnabled,
+            },
+            {
+              target: "check_token",
+              actions: assignFrogState,
+              cond: (_, event) => event.data.canMint,
+            },
+            {
+              target: "finished",
+              actions: assignFrogState,
+            },
+          ],
+          onError: {
+            target: "error",
+            actions: assignErrorMessage,
+          },
+        },
+      },
+      check_whitelist: {
+        invoke: {
+          src: async () => {
+            const isWalletWhitelisted = await metamask
+              .getFrog()
+              .isWalletWhitelisted();
+            console.log("is wallet whitelisted?", isWalletWhitelisted);
+
+            return { isWalletWhitelisted };
+          },
+          onDone: [
+            {
+              target: "check_token",
+              cond: (_, event) => event.data.isWalletWhitelisted,
+            },
+            {
+              target: "not_whitelisted",
+            },
+          ],
+          onError: {
+            target: "error",
+            actions: assignErrorMessage,
+          },
+        },
+      },
+      check_token: {
+        invoke: {
+          src: async () => {
+            const isTokenApproved = await metamask
+              .getToken()
+              .isTokenApprovedForFrogMint();
+            console.log("is token approved?", isTokenApproved);
+
+            return { isTokenApproved };
+          },
+          onDone: [
+            {
+              target: "mint",
+              cond: (_, event) => event.data.isTokenApproved,
+            },
+            {
+              target: "approve",
+            },
+          ],
+          onError: {
+            target: "error",
+            actions: assignErrorMessage,
+          },
+        },
+      },
+      approve: {
+        on: {
+          APPROVE: {
+            target: "approving",
+          },
+        },
+      },
+      approving: {
+        invoke: {
+          src: async () => {
+            const _approve = await approve(frogAddress);
+            console.log(_approve);
+
+            return { _approve };
+          },
+          onDone: [
+            {
+              target: "check_token",
+            },
+          ],
+          onError: {
+            target: "error",
+            actions: assignErrorMessage,
+          },
+        },
+      },
+      mint: {
+        on: {
+          MINT: {
+            target: "minting",
+          },
+        },
+      },
+      minting: {
+        invoke: {
+          src: async () => {
+            const farm = await metamask.getFarm()?.getFarms();
+            const mint = await mintFrog({ farmId: Number(farm[0].tokenId) });
+            console.log(mint);
+
+            return { mint };
+          },
+          onDone: [
```

Comment:
based on `mintFrog()`, `verified` always return `true` so i think there's no need for `cond`. also can omit `[]` here in `onDone`

Reply:
You're right, will refactor this thanks!